Examples/Io/Performance/ActsExamples/Io/Performance/VertexPerformanceWriter.cpp

```diff
@@ -402,89 +461,220 @@ ActsExamples::ProcessCode ActsExamples::VertexPerformanceWriter::writeT(
+            if (variance <= 0) {
+              std::string tempStr;
+              if (afterFit) {
+                tempStr = "after";
+              } else {
+                tempStr = "before";
+              }
+              ACTS_WARNING("Nonpositive variance "
+                           << tempStr << " vertex fit: Var(" << variableStr
+                           << ") = " << variance << " <= 0.");
               return std::numeric_limits<double>::quiet_NaN();
             }
-            return (vtx.fullPosition()[i] - truePos[i]) / std;
+            double std = std::sqrt(variance);
+            return diff / std;
           };
 
-          m_pullX.push_back(pull(Acts::FreeIndices::eFreePos0));
-          m_pullY.push_back(pull(Acts::FreeIndices::eFreePos1));
-          m_pullZ.push_back(pull(Acts::FreeIndices::eFreePos2));
-          m_pullT.push_back(pull(Acts::FreeIndices::eFreeTime));
-
-          m_covXX.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos0,
-                                                 Acts::FreeIndices::eFreePos0));
-          m_covYY.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos1,
-                                                 Acts::FreeIndices::eFreePos1));
-          m_covZZ.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos2,
-                                                 Acts::FreeIndices::eFreePos2));
-          m_covTT.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreeTime,
-                                                 Acts::FreeIndices::eFreeTime));
-          m_covXY.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos0,
-                                                 Acts::FreeIndices::eFreePos1));
-          m_covXZ.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos0,
-                                                 Acts::FreeIndices::eFreePos2));
-          m_covXT.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos0,
-                                                 Acts::FreeIndices::eFreeTime));
-          m_covYZ.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos1,
-                                                 Acts::FreeIndices::eFreePos2));
-          m_covYT.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos1,
-                                                 Acts::FreeIndices::eFreeTime));
-          m_covZT.push_back(vtx.fullCovariance()(Acts::FreeIndices::eFreePos2,
-                                                 Acts::FreeIndices::eFreeTime));
-
-          m_nTracksOnTruthVertex.push_back(nTracksOnTruthVertex);
-          m_nTracksOnRecoVertex.push_back(tracks.size());
-
-          m_trackVtxMatchFraction.push_back(trackVtxMatchFraction);
-
-          // Next vertex now
-          break;
+          const Acts::ActsVector<4>& truePos = particle.fourPosition();
+          // Save reconstructed/true vertex position only in the first iteration
+          // to avoid duplicates
+          if (count == 0) {
```

Comment:
I don't fully understand why this is done?
Is it because the vertex outputs are vectors, and you have multiple tracks per vector? Would it not be cleaner to use vectors-of-vectors for the track quantities, and keep the outer vector sized the same?